rainier-core/src/main/scala/com/stripe/rainier/core/SBCModel.scala

```diff
@@ -0,0 +1,48 @@
+package com.stripe.rainier.core
+
+import com.stripe.rainier.compute._
+import com.stripe.rainier.sampler._
+
+trait SBCModel[L, T] {
+  implicit val rng: RNG = ScalaRNG(1528673302081L)
+  def sbc: SBC[L, T]
+  def main(args: Array[String]): Unit = sbc.animate(HMC(1), 10000, 1000)
+  def samples: List[T] = sbc.posteriorSamples(goldset.size)
+  def goldset: List[T]
+  def description: String
+}
+
+object SBCUniformNormal extends SBCModel[Continuous, Double] {
+  def sbc = SBC(Uniform(0, 1))((x: Real) => Normal(x, 1))
+  def goldset =
+    List(0.770217434378013, 1.2877991075049935, 1.0761575302777342,
+      0.06063426627950119, 2.235772959187611, 1.5916139704386643,
+      0.4680530806057483, -1.3360844052727425, 0.6558428534558145,
+      0.4233930395976211)
+  def description = "Normal(x,1) with Uniform(0,1) prior"
```

Comment:
everywhere we have `def` here we could use `val`

Reply:
I changed everything to a `val` and got ```[info] Running com.stripe.rainier.core.SBCUniformNormal
[error] (run-main-1) java.lang.ExceptionInInitializerError
[error] java.lang.ExceptionInInitializerError
[error] 	at com.stripe.rainier.core.SBCUniformNormal.main(SBCModel.scala)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.lang.reflect.Method.invoke(Method.java:498)
[error] 	at sbt.Run.invokeMain(Run.scala:93)
[error] 	at sbt.Run.run0(Run.scala:87)
[error] 	at sbt.Run.execute$1(Run.scala:65)
[error] 	at sbt.Run.$anonfun$run$4(Run.scala:77)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] 	at sbt.util.InterfaceUtil$$anon$1.get(InterfaceUtil.scala:10)
[error] 	at sbt.TrapExit$App.run(TrapExit.scala:252)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] Caused by: scala.UninitializedFieldError: Uninitialized field: /Users/mio/stripe/rainier/rainier-tests/src/main/scala/com/stripe/rainier/core/SBCModel.scala: 29```
then changed two things back and it worked. Seems like the things that are no implemented in the trait have to be defs and the rest can be vals.

Reply:
Ah, this is about initialization order - you need to make sure the vals are in dependency order from top to bottom (or make them `def` or `lazy val`).